### DISCUSSION TITLE
Return if expression is not null

### DISCUSSION THREAD

**Discussion by simongoldstone** [dotnet/csharplang#6245] [Language Ideas]
I feel a useful language feature would be if we could return from a method only if the expression being tested is not null. For example, instead of:

```
foreach (var item in items)
{
    var match = item.Match("str");
    if (match != null)
        return match;
}

return defaultMatch;
```

We could introduce something like:

```
foreach (var item in items)
    return? item.Match("str");

return defaultMatch;
```

In this example, **return?** only returns if the expression is not null.

Removes the need for braces and a temporary variable.

**Comment by iam3yal** [comment]
There is a proposal to get `return`, `continue`, `break` and iirc `goto` as expressions so you could write something like this:
```csharp
item.Match("str") is MyMatchType match ? return match : continue;
```
Not sure whether this will ever happen.

You can write something like the following today although personally I prefer to use braces for single statements too.
```csharp
foreach (var item in items) 
    if (item.Match("str") is MyMatchType match) return match;
```

As a side note I think the syntax `return?` is confusing.

**Reply by simongoldstone** [reply]
I agree return? may be confusing, but it is similar at least to object?.Property.
Thanks for the heads up on the existing proposal.

**Comment by VWilcox2000** [comment]
I agree on return? being slightly on the confusing side, though "return if" the intuitive replacement doesn't feel like c#.

Personally, I almost NEVER have more than a single return at end of any function, but would use such a short-hand option on yield return.  For example

        foreach (object item in items)
              yield return? item as MyMatchType;

or with code already under consideration you mentioned:

       foreach (object item in items)
             item is MyMatchType item ? yield return item : continue; 

Though if that was a pattern, I would think the option to omit : would be useful in this pattern.

**Reply by simongoldstone** [reply]
I agree re return? could perhaps be confusing. Regarding a single return in a function, it's useful in cases where you don't want several nested if statements.

**Reply by Rekkonnect** [reply]
> Personally, I almost NEVER have more than a single return at end of any function, but would use such a short-hand option on yield return.

`yield return?` is #5063, and I'm in favor of `return?` too, but it's just slightly less common

**Comment by BreyerW** [comment]
referencing proposal about return, continue etc. as expression #867

note it is in Likely never category but in the middle of 2021 LDM changed mind and decided to look at it again in future but did not update category so disregard that

**Reply by simongoldstone** [reply]
Thanks for update
